Add more documentation along with minor maintenance (#166)
jenkinsci/credentials-plugin: docs/faq.adoc

== FAQs

This document provides a quick reference for frequently questioned concepts and features in Credentials.

=== Where do credentials get stored?

By default, Jenkins uses an internal credentials store where credentials secret data are encrypted at rest in the Jenkins controller file system.
This security model does not prevent anyone with read access to the controller file system from reading and decrypting those secrets.
For more sensitive secrets, it is highly recommended to install and use an external credentials provider plugin which delegates the secure storage of secrets to a proper key management system.

> Various cloud providers have secrets management and HSMs available for this purpose, and open source software such as Vault or Conjur provide more generic secrets management for multiple platforms.

Reviewer: Maybe mention `kubernetes-credentials-provider` here as well.

=== How should I organize my credentials?

It's important to understand the access model of credentials in order to organize them most appropriately.
Credentials are accessed through a combination of scopes and contexts, which are further categorized into types and domains.

The _context_ of a credential is the path in Jenkins where the credentials are accessed.

> Contexts include the root Jenkins context, jobs, users, agents, views, folders, and generally any other item type.

Reviewer: ```suggestion
Contexts include the root Jenkins context, jobs, users, agents, views, folders, and generally any other kind of model object.
```

since _item_ refers to folders and jobs but not views or agents for example.

System-scoped credentials are only visible in the root Jenkins context (typically useful for things like agent SSH keys).
User-scoped credentials are only visible to a context running as that user (such as a user-provided credentials parameter to a build).
Global-scoped credentials are visible to their associated contexts and any children contexts; this is the default credential scope.

[IMPORTANT]
====
Global scoped credentials in the Jenkins Credentials Provider (system credentials) are accessible to any job in the system!
====

Credentials _stores_ are the underlying storage mechanism for secrets, and credentials _providers_ combine them with the functionality to manage secrets in a context for a certain set of scopes.
For example, the Jenkins Credentials Provider uses a system-wide credentials store at the root context and supports both system and global scoped credentials.
Folder plugins provide a Folder Credentials Provider to anchor credentials storage in a folder along with the advanced authorization strategies possible there.
User-scoped credentials are primarily accessed through the User Credentials Provider which stores secrets with the user profile in Jenkins.

> Credentials _types_ include things like username/password, secret text, SSH key, API token, etc.

Reviewer: Generally either 
```suggestion
Credentials _types_ include things like username/password, secret text, SSH key, etc.
```

Tokens are generally represented as either passwords with an arbitrary username, or secret text.

> Finally, credentials _domains_ provide a way to help organize credentials by domain name, URI, protocol, and other requirements which can make it much easier to identify the appropriate credentials to use from the UI.

Reviewer: ```suggestion
Finally, credentials _domains_ provide a way to help organize credentials by domain name, URI, protocol, and other requirements which can make it much easier to identify the appropriate credentials to use from the UI.
(Domains are _not_ enforced at usage time, only for purposes of the UI.)
```

=== What plugins do I need to be secure?

First, ensure that you are following the https://www.jenkins.io/doc/book/system-administration/security/[Jenkins security best practices].
In particular, you'll need to use a non-trivial authorization strategy such as https://plugins.jenkins.io/matrix-auth/[matrix], https://plugins.jenkins.io/role-strategy/[role-based], or https://plugins.jenkins.io/folder-auth/[folder-based]; the trivial authorization strategies would defeat the purpose of securing your secrets!

> Next, you'll need the https://plugins.jenkins.io/authorize-project/[Authorize Project] plugin so that more fine-grained permission controls are available to configure to jobs so that credentials accessible to one job aren't necessarily accessible to every other job.

Reviewer: `authorize-project` is not necessary for that purpose; you can use folder-scoped credentials.

Author: That's true, though folder-stored (there's no folder scope; there's way too many concepts in this plugin) credentials are a little coarser than what `authorize-project` offers. I'll update this to clarify, though, thanks.

The https://plugins.jenkins.io/ownership/[Job and Node ownership] plugin extends many of the mentioned authorization strategies to allow for further configuration.
Additional hardening can be applied via the https://plugins.jenkins.io/job-restrictions/[Job Restrictions] plugin.

> When working with especially sensitive secrets, use of a hardware security module (HSM) is recommended for physical storage and protection of secret and private keys.

Reviewer: Not sure how relevant this is in the context of Jenkins credentials.

Author: The relevance is that users should not expect HSM-level security with this credentials plugin despite looking super fancy. For example, suppose you want to automate code signing with a TLS certificate. You might want extra hardware security for the signing key so that nothing else can use it. For Jenkins, we use the Azure equivalent of Vault to do that, for example.

Reviewer: Maybe rephrase to clarify that if you want HSM or similar levels of security, you should not be using Jenkins credentials. Currently it makes it sound like there is some vaguely described means of using an HSM to help lock down Jenkins credentials.

Author: Good idea :)

> == FAQs

Reviewer 2: ```suggestion
== FAQ
```

Per https://en.wikipedia.org/wiki/FAQ

This document provides a quick reference for frequently questioned concepts and features in Credentials.

=== Where do credentials get stored?

By default, Jenkins uses an internal credentials store where credentials secret data are encrypted at rest in the Jenkins controller file system.
This security model does not prevent anyone with read access to the controller file system from reading and decrypting those secrets.
For more sensitive secrets, it is highly recommended to install and use an external credentials provider plugin which delegates the secure storage of secrets to a proper key management system.

> Various cloud providers (including https://plugins.jenkins.io/aws-secrets-manager-credentials-provider/[AWS], https://plugins.jenkins.io/azure-keyvault/[Azure], and https://plugins.jenkins.io/gcp-secrets-manager-credentials-provider/[GCP]) have secrets management and HSMs available for this purpose, and open source software such as https://plugins.jenkins.io/kubernetes-credentials-provider/[Kubernetes], https://plugins.jenkins.io/hashicorp-vault-plugin/[Vault], or https://plugins.jenkins.io/conjur-credentials/[Conjur] provide more generic secrets management for multiple platforms.

Reviewer 2: Prefer a list format.

Also, it might be useful to link (with disclaimer about technical content) to the generated list at https://www.jenkins.io/doc/developer/extensions/credentials/#credentialsprovider ?

Author: That link is misleading as some of the external secrets manager plugins use a hybrid approach where secrets are stored in the service, but Jenkins stores credentials internally which are basically placeholders for the external credentials. Due to the various ways that credentials plugins can be made, it seems as though not everyone has used the same approach for this.

There are several solutions available for storing and securing secrets, many of which integrate with hardware security modules (HSMs) for additional physical security of secrets.
Examples of common external credentials stores for cloud providers or self-hosted environments include:

- https://plugins.jenkins.io/aws-secrets-manager-credentials-provider/[AWS]
- https://plugins.jenkins.io/azure-keyvault/[Azure]
- https://plugins.jenkins.io/conjur-credentials/[Conjur]
- https://plugins.jenkins.io/gcp-secrets-manager-credentials-provider/[GCP]
- https://plugins.jenkins.io/kubernetes-credentials-provider/[Kubernetes]
- https://plugins.jenkins.io/hashicorp-vault-plugin/[Vault]

A full listing (including some developer-level details) of https://www.jenkins.io/doc/developer/extensions/credentials/[plugins that provide credentials extensions] is available in the developer documentation.

=== How should I organize my credentials?

It's important to understand the access model of credentials in order to organize them most appropriately.

> Credentials are accessed through a combination of scopes and contexts, which are further categorized into types and domains.

Author: There's already verbose descriptions of these things in the user guide, but said guide is enormous. Considering barely anyone can get the terminology consistent, I figured explaining it somewhere in the FAQ would help. Perhaps the concepts can be split into "what is _foo_?"

> Credentials are accessed through a combination of scopes and contexts, which are further categorized into types and domains.

Reviewer 3: I'll bet that would work better. I like the idea of getting these basic concepts more accessible to users, but they're still kind of buried inside something else.
